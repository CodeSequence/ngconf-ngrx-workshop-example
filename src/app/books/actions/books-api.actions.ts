import { Book } from "src/app/shared/models/book.model";
import { Action } from "@ngrx/store";
