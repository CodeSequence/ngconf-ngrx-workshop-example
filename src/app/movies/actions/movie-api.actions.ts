import { createAction, props } from "@ngrx/store";
import { Movie } from "src/app/shared/models/movie.model";

export const loadMoviesSuccess = createAction(
  "[Movie API] Load Movies Success",
  props<{ movies: Movie[] }>()
);

export const loadMoviesFailure = createAction(
  "[Movie API] Load Movies Failure"
);

export const createMovieSuccess = createAction(
  "[Movie API] Create Movie Success",
  props<{ movie: Movie }>()
);

export const createMovieFailure = createAction(
  "[Movie API] Create Movie Failure"
);

export const updateMovieSuccess = createAction(
  "[Movie API] Update Movie Success",
  props<{ movie: Movie }>()
);

export const updateMovieFailure = createAction(
  "[Movie API] Update Movie Failure",
  props<{ movie: Movie }>()
);

export const deleteMovieSuccess = createAction(
  "[Movie API] Delete Movie Success",
  props<{ movieId: string }>()
);

export const deleteMovieFailure = createAction(
  "[Movie API] Delete Movie Failure",
  props<{ movie: Movie }>()
);

export type Union = ReturnType<
  | typeof loadMoviesSuccess
  | typeof loadMoviesFailure
  | typeof createMovieSuccess
  | typeof createMovieFailure
  | typeof updateMovieSuccess
  | typeof updateMovieFailure
  | typeof deleteMovieSuccess
  | typeof deleteMovieFailure
>;
