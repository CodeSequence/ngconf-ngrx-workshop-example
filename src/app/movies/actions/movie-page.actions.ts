import { createAction, props } from "@ngrx/store";
import { MovieRequiredProps, Movie } from "src/app/shared/models/movie.model";

export const enter = createAction("[Movie Page] Enter");

export const selectMovie = createAction(
  "[Movie Page] Select Movie",
  props<{ movieId: string }>()
);

export const createMovie = createAction(
  "[Movie Page] Create Movie",
  props<{ movie: MovieRequiredProps }>()
);

export const updateMovie = createAction(
  "[Movie Page] Update Movie",
  props<{ movie: Movie; changes: MovieRequiredProps }>()
);

export const deleteMovie = createAction(
  "[Movie Page] Delete Movie",
  props<{ movie: Movie }>()
);

export type Union = ReturnType<
  | typeof enter
  | typeof selectMovie
  | typeof createMovie
  | typeof updateMovie
  | typeof deleteMovie
>;
