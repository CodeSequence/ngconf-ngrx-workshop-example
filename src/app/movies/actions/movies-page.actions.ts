import { createAction, props } from "@ngrx/store";
import { MovieRequiredProps, Movie } from "src/app/shared/models/movie.model";

export const enter = createAction("[Movies Page] Enter");

export const selectMovie = createAction(
  "[Movies Page] Select Movie",
  props<{ movieId: string }>()
);

export const clearSelectedMovie = createAction(
  "[Movies Page] Clear Selected Movie"
);

export const createMovie = createAction(
  "[Movies Page] Create Movie",
  props<{ movie: MovieRequiredProps }>()
);

export const updateMovie = createAction(
  "[Movies Page] Update Movie",
  props<{ movie: Movie; changes: MovieRequiredProps }>()
);

export const deleteMovie = createAction(
  "[Movies Page] Delete Movie",
  props<{ movie: Movie }>()
);

export type Union = ReturnType<
  | typeof enter
  | typeof selectMovie
  | typeof clearSelectedMovie
  | typeof createMovie
  | typeof updateMovie
  | typeof deleteMovie
>;
