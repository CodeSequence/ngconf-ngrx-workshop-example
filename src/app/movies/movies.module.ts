import { NgModule } from "@angular/core";
import { EffectsModule } from "@ngrx/effects";
import { MovieApiEffects } from "./movie-api.effects";

@NgModule({ imports: [EffectsModule.forFeature([MovieApiEffects])] })
export class MoviesModule {}
