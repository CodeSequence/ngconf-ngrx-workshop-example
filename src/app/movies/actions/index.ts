import * as MovieApiActions from "./movie-api.actions";
import * as MoviesPageActions from "./movies-page.actions";

export { MovieApiActions, MoviesPageActions };
