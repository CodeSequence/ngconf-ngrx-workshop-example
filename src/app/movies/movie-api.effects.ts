import { Injectable } from "@angular/core";
import { Effect, Actions, ofType } from "@ngrx/effects";
import { of } from "rxjs";
import { mergeMap, map, catchError, concatMap } from "rxjs/operators";
import { MovieApiActions, MoviePageActions } from "./actions";
import { MovieService } from "../shared/services/movie.service";

@Injectable()
export class MovieApiEffects {
  constructor(
    private actions$: Actions<MoviePageActions.Union>,
    private movieService: MovieService
  ) {}

  @Effect() createMovie$ = this.actions$.pipe(
    ofType(MoviePageActions.createMovie.type),
    mergeMap(action =>
      this.movieService.create(action.movie).pipe(
        map(movie => MovieApiActions.createMovieSuccess({ movie })),
        catchError(() => of(MovieApiActions.createMovieFailure()))
      )
    )
  );

  @Effect() updateMovie$ = this.actions$.pipe(
    ofType(MoviePageActions.updateMovie.type),
    concatMap(action =>
      this.movieService.update(action.movie.id, action.changes).pipe(
        map(movie => MovieApiActions.updateMovieSuccess({ movie })),
        catchError(() =>
          of(MovieApiActions.updateMovieFailure({ movie: action.movie }))
        )
      )
    )
  );

  @Effect() deleteMovie$ = this.actions$.pipe(
    ofType(MoviePageActions.deleteMovie.type),
    mergeMap(action =>
      this.movieService.delete(action.movie.id).pipe(
        map(() =>
          MovieApiActions.deleteMovieSuccess({ movieId: action.movie.id })
        ),
        catchError(() =>
          of(MovieApiActions.deleteMovieFailure({ movie: action.movie }))
        )
      )
    )
  );
}
