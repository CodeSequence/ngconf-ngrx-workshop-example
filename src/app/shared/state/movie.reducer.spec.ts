import { MovieApiActions, MoviePageActions } from "src/app/movies/actions";
import { Movie } from "../models/movie.model";
import { reducer, initialState } from "./movie.reducer";

describe("Movie Reducer", () => {
  it("should return the initial state when initialized", () => {
    const state = reducer(undefined, { type: "@@init" } as any);

    expect(state).toBe(initialState);
  });

  it("should load all movies when the API loads them all successfully", () => {
    const movies: Movie[] = [{ id: "1", name: "Green Lantern", rating: 0 }];
    const action = MovieApiActions.loadMoviesSuccess({ movies });

    const state = reducer(initialState, action);

    expect(state).toMatchSnapshot();
  });

  it("should add newly created movies to the state", () => {
    const movie: Movie = { id: "1", name: "Arrival", rating: 4 };
    const action = MovieApiActions.createMovieSuccess({ movie });

    const state = reducer(initialState, action);

    expect(state).toMatchSnapshot();
  });

  it("should remove movies from the state when they are deleted", () => {
    const movie: Movie = { id: "1", name: "mother!", rating: 1.5 };
    const firstAction = MovieApiActions.createMovieSuccess({ movie });
    const secondAction = MoviePageActions.deleteMovie({ movie });

    const state = [firstAction, secondAction].reduce(reducer, initialState);

    expect(state).toMatchSnapshot();
  });

  it("should roll back a deletion if deleting a movie fails", () => {
    const movie: Movie = { id: "1", name: "Black Panther", rating: 4 };
    const firstAction = MovieApiActions.createMovieSuccess({ movie });
    const secondAction = MoviePageActions.deleteMovie({ movie });
    const thirdAction = MovieApiActions.deleteMovieFailure({ movie });

    const state = [firstAction, secondAction, thirdAction].reduce(
      reducer,
      initialState
    );

    expect(state).toMatchSnapshot();
  });

  it("should apply changes to a movie when a movie is updated", () => {
    const movie: Movie = { id: "1", name: "Blade Runner", rating: 2 };
    const changes = { name: "Blade Runner (Final Cut)", rating: 4.5 };
    const firstAction = MovieApiActions.createMovieSuccess({ movie });
    const secondAction = MoviePageActions.updateMovie({ movie, changes });

    const state = [firstAction, secondAction].reduce(reducer, initialState);

    expect(state).toMatchSnapshot();
  });

  it("should rollback changes to a movie if there is an error when updating it with the API", () => {
    const movie: Movie = {
      id: "1",
      name: "Star Wars: A New Hope",
      rating: 3.5
    };
    const changes = {
      name: "Star Wars: A New Hope (Special Edition)",
      rating: 2.5
    };
    const firstAction = MovieApiActions.createMovieSuccess({ movie });
    const secondAction = MoviePageActions.updateMovie({ movie, changes });
    const thirdAction = MovieApiActions.updateMovieFailure({ movie });

    const state = [firstAction, secondAction, thirdAction].reduce(
      reducer,
      initialState
    );

    expect(state).toMatchSnapshot();
  });
});
