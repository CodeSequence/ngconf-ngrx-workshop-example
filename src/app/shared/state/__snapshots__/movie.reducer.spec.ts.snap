// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie Reducer should add newly created movies to the state 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "id": "1",
      "name": "Arrival",
      "rating": 4,
    },
  },
  "ids": Array [
    "1",
  ],
}
`;

exports[`Movie Reducer should apply changes to a movie when a movie is updated 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "id": "1",
      "name": "Blade Runner (Final Cut)",
      "rating": 4.5,
    },
  },
  "ids": Array [
    "1",
  ],
}
`;

exports[`Movie Reducer should load all movies when the API loads them all successfully 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "id": "1",
      "name": "Green Lantern",
      "rating": 0,
    },
  },
  "ids": Array [
    "1",
  ],
}
`;

exports[`Movie Reducer should remove movies from the state when they are deleted 1`] = `
Object {
  "entities": Object {},
  "ids": Array [],
}
`;

exports[`Movie Reducer should roll back a deletion if deleting a movie fails 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "id": "1",
      "name": "Black Panther",
      "rating": 4,
    },
  },
  "ids": Array [
    "1",
  ],
}
`;

exports[`Movie Reducer should rollback changes to a movie if there is an error when updating it with the API 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "id": "1",
      "name": "Star Wars: A New Hope",
      "rating": 3.5,
    },
  },
  "ids": Array [
    "1",
  ],
}
`;
